### PR TITLE
Fix container errors not being sent via pod removal API

### LINKED ISSUE #19159: New podman pod remove API error message is less detailed
### Issue Description

In 4.6.0-rc1 the `libpods/pod` DELETE endpoint's error message changed to:

image

From:

image

The new error message does not include anything about it being unable delete running/paused containers.

### Steps to reproduce the issue

Steps to reproduce the issue
```
[root@fedora-testing-127-0-0-2-2201 ~]# podman pod ps
POD ID        NAME           STATUS      CREATED         INFRA ID      # OF CONTAINERS
64ec75433156  brave_solomon  Running     24 minutes ago  c168e97d7e68  2
[root@fedora-testing-127-0-0-2-2201 ~]# curl -X DELETE -s -g --no-buffer --unix-socket /run/podman/podman.sock 'http://localhost/v1.12/libpod/pods/64ec75433156'
{"cause":"removing pod containers","message":"not all containers could be removed from pod 64ec754331564082b0ccbf078268d701ad009a415ad1dfdaca64425cbe8d8821: removing pod containers","response":500}
```

### Describe the results you received

See above :)

### Describe the results you expected

I'm wondering if the error should inform that the container is running/paused so requires force to be removed.

### Podman in a container

No

### Privileged Or Rootless

None

### Upstream Latest Release

Yes

### Additional environment details

Additional environment details

### Additional information

Additional information like issue happens only occasionally or issue happens with a particular architecture or on a particular setting

### PR DESCRIPTION
**Pull request by mheon**
When I reworked pod removal to provide more detailed errors (including per-container errors, not just a single multierror with all errors squashed), I made it part of the struct returned by the REST API and assumed that would be enough to get errors through to clients. Unfortunately, in case of an overarching error removing the pod (as any error with any container would cause), we don't send the response struct that would include the container errors - we just send a standardized REST error. We could work around this with custom, potentially backwards incompatible error handling for the REST pod delete endpoint, or we could just do what was done before, and package up all the errors in a multierror to send to the other side. Of those options, the multierror seems far simpler.

Fixes #19159

```release-note
NONE
```
(Bug never appeared in a published non-RC release, so no need to document it)

### CODE REVIEW

**Comment by mheon** [comment]
Needs a test, will write one tomorrow morning

**Review by Luap99 — COMMENTED** [review]
LGTM, I prefer that over sending errors in an extra type.
However I think one problem is that remote error will look different than local ones which makes testing harder.

**Comment by edsantiago** [comment]
To run tests locally:
```console
$ [optional sudo] ./test/apiv2/test-apiv2 40
```

**Comment by mheon** [comment]
Tests should pass now

File: test/apiv2/40-pods.at [hunk 1]
```diff
@@ -52,11 +61,16 @@ t POST "libpod/pods/fakename/unpause" 404\
   .cause="no such pod" \
   .message="no pod with name or ID fakename found: no such pod"
 
+t DELETE libpod/pods/foo?force=false 500 \
+  .cause~".*cannot remove container .* as it is running.*"
```
Comment (edsantiago) [inline test/apiv2/40-pods.at:65]: Playing around some more, it looks like `message` and `cause` are identical. If that is the intention, you may want to duplicate line 64, `s/cause/message/`. (Backslashing line 64 of course). You may also want to add as much or little other context as you wish, e.g.
```
   .cause~"not all containers could be removed from pod.*cannot remove container .* as it is running - running or paused containers blah blah blah you get my point"
```
(Might as well wait to see how CI goes though)

**Comment by TomSweeneyRedHat** [comment]
Other than @edsantiago 's comment, LGTM once the tests are hip.

File: test/apiv2/40-pods.at [hunk 2]
```diff
@@ -52,11 +61,16 @@ t POST "libpod/pods/fakename/unpause" 404\
...
 
 t POST libpod/pods/foo/stop 200 \
   .Errs=null \
   .Id=$pod_id
 
+# Remove container as it is no longer required
+t DELETE libpod/containers/$cid 200 .[0].Id=$cid
+
```
Comment (edsantiago) [inline test/apiv2/40-pods.at:73]: Hey, this is out of scope, and please ONLY if you need to repush... but... line 102 below, `sleep 999`, could you change it to `top` instead?

Reason: way way down, line 120, `DELETE /the/pod`, takes ten seconds because `sleep` is not interruptible. I tried adding `\&t=0` but that seems to be ignored by `pod delete`. Thank you!

**Comment by edsantiago** [comment]
Hey, should you need to repush for any reason, may I suggest one more diff?
```diff
diff --git a/test/system/200-pod.bats b/test/system/200-pod.bats
index 442517b46..f159ed6ee 100644
--- a/test/system/200-pod.bats
+++ b/test/system/200-pod.bats
@@ -55,6 +55,16 @@ function teardown() {
         is "$output" ".*0 \+1 \+0 \+[0-9. ?s]\+/pause" "there is a /pause container"
     fi
 
+    # Cannot remove pod while containers are still running. Error messages
+    # differ slightly between local and remote; these are the common elements.
+    run_podman 125 pod rm $podid
+    assert "${lines[0]}" =~ "Error: not all containers could be removed from pod $podid: removing pod containers.*" \
+           "pod rm while busy: error message line 1 of 3"
+    assert "${lines[1]}" =~ "cannot remove container .* as it is running - running or paused containers cannot be removed without force: container state improper" \
+           "pod rm while busy: error message line 2 of 3"
+    assert "${lines[2]}" =~ "cannot remove container .* as it is running - running or paused containers cannot be removed without force: container state improper" \
+           "pod rm while busy: error message line 3 of 3"
+
     # Clean up
     run_podman --noout pod rm -f -t 0 $podid
     is "$output" "" "output should be empty"
```
Confirmed to fail on `main`, pass on your PR.

File: pkg/api/handlers/libpod/pods.go [hunk 3]
```diff
@@ -248,6 +249,21 @@ func PodDelete(w http.ResponseWriter, r *http.Request) {
...
+			// special error-handling for just this endpoint: use a
+			// multierror to package up all container errors.
+			var allCtrErrors error
+			for _, ctrErr := range ctrs {
+				allCtrErrors = multierror.Append(allCtrErrors, ctrErr)
+			}
+
+			// This loses error typing, but since it's going to be
+			// JSON text once it hits the wire, it doesn't matter.
+			err = fmt.Errorf("%s. %s", err.Error(), allCtrErrors.Error())
```
Comment (vrothberg) [inline pkg/api/handlers/libpod/pods.go:264]: Alternatively, you could use `pkg/errorhandling.JoinErrors(...)`.

File: test/apiv2/40-pods.at (hunk 2 above)
Comment (mheon) [reply test/apiv2/40-pods.at:73]: Done

File: pkg/api/handlers/libpod/pods.go (hunk 3 above)
Comment (mheon) [reply pkg/api/handlers/libpod/pods.go:264]: That'll nest a second multierror on top of the first one, I don't think that's a good idea, formatting wise

**Comment by mheon** [comment]
Force-pushed with updates, all Ed's test suggestions should be included now.
